src/v/ssx/abort_source.h

```diff
@@ -80,4 +81,56 @@ class sharded_abort_source {
 struct shutdown_requested_exception : ss::abort_requested_exception {};
 struct connection_aborted_exception : ss::abort_requested_exception {};
 
+/// A composite abort source is a new abort source that is subscribed to
+/// the given abort sources and will request abort if any of them
+/// requests abort.
+///
+/// The abort on the composite abort source does not propagate to the
+/// underlying abort sources, it only requests abort on the composite
+/// abort source itself. Pay attention to this as it changes the
+/// semantics of your code if you rely on propagating aborts to the caller.
+class composite_abort_source {
+public:
+    composite_abort_source(
+      ss::abort_source& as1, ss::abort_source& as2) noexcept {
+        for (auto as : {&as1, &as2}) {
+            if (as->abort_requested()) {
+                _as.request_abort_ex(as->abort_requested_exception_ptr());
+                return;
+            }
+        }
+
+        for (auto as : {&as1, &as2}) {
+            // Safe to capture `this` here because the subscription lambda
+            // will be destroyed together with the `composite_abort_source`
+            // instance.
+            auto sub = as->subscribe(
+              [this](const std::optional<std::exception_ptr>& ex) noexcept {
+                  if (ex) {
+                      _as.request_abort_ex(*ex);
+                  } else {
+                      _as.request_abort();
+                  }
+              });
+
+            vassert(
+              sub, "Empty subscription after we checked for abort_requested");
+
+            _subs.emplace_back(std::move(*sub));
+        }
+    }
+
+    /// Returns the underlying abort source for interoperability with
+    /// the Seastar APIs.
+    ss::abort_source& as() noexcept { return _as; }
+
+    /// Returns the underlying abort source for interoperability with
+    /// the Seastar APIs.
+    const ss::abort_source& as() const noexcept { return _as; }
```

Comment:
it would be neat to be able to expose a limited interface or abort source wrapper that didn't allow requesting abort so that the api disallowed uses which look like propogating aborts downward instead of upward (the allowed behavior).